### ISSUE TITLE
Async .net side waiter for bound objects

### ISSUE THREAD

**Issue by mitchcapper** [cefsharp/CefSharp#2362]
I mentioned this briefly previously but I think it might be of value to have a .net side task that you could await for an object to be bound into existence. This can be partially accomplished with the nice ObjectBoundInJavascript event but as it doesn't fire if already bound it cannot be done as easily.

This may make it easier for people to move from the old binding to the new binding system if they only 

use bound objects to exfil data from JS and not get results of methods.

It allows for something like:
```
helper = new Helper(browser);
browser.JavascriptObjectRepository.Register("Tester", new TestClass(), true, CefSharp.BindingOptions.DefaultBinder);
await helper.EnsureRegistered("Tester");
browser.ExecuteScriptAsync("Tester.test('hi there');");
```

Attached is a .net helper class that emulates the process that I could see being used.  It has added complexity as we cannot tell from .net if an object is already bound (I guess could call EvaluateScriptAsync to test each window object).  This could be resolved with another event handler of ObjectAlreadyBoundInJavascript called for any time bindobjectasync is called with an already bound object.

[Helper.cs](https://github.com/cefsharp/CefSharp/files/1910669/Helper.cs.txt)

Let me know thoughts or if worthwhile.  This helper works for me (and actually could be simplified to just wait for BindObjectAsync rather than resolving after each bound object).  Part of the current flow is to show what would be possible with the additional event handler, just one line of code with no helper class bound into the JS namespace.   It could be taken a step further with no JS injection if the bind call was possible from .net code.

**Comment by amaitland** [comment]
> This can be partially accomplished with the nice ObjectBoundInJavascript event but as it doesn't fire if already bound it cannot be done as easily.

Adding a custom set of `EventArgs` for `ObjectBoundInJavascript` and adding some extra properties like `bool AlreadyBound` and `bool IsCached` would be pretty trivial. The question I have does it make sense to do so? Would be be confusing in any way?

I was originally tossing up between `ObjectBoundInJavascript` and `ObjectsBoundInJavascript` with the latter being called a single time with a list of all bound objects, maybe it's worth adding a second event with the different behavior? 

> Part of the current flow is to show what would be possible with the additional event handler, just one line of code with no helper class bound into the JS namespace.

Do you think a totally new event specially for this purpose would be preferable? If so what would it be called?

> It could be taken a step further with no JS injection if the bind call was possible from .net code.

My initial though on automatically binding objects upon context creation were that it's not particularly helpful when `Javascript` was driving the interaction. The simplest possible option that should work currently would be implementing `IRenderProcessMessageHandler` and using some code similar to:

```c#
void IRenderProcessMessageHandler.OnContextCreated(IWebBrowser browserControl, IBrowser browser, IFrame frame)
{
	frame.ExecuteJavaScriptAsync("CefSharp.BindObjectAsync();");
}
```

The advantage of executing the code in the `IRenderProcessMessageHandler` is you get more control, only bind for certain frames, etc. Adding an option to the render process it's self would be an more than likely a simple `bool`, I was trying to avoid that as it's not very user friendly in my opinion.

Probably need a semi real work use case to make sure whatever option choose is easy enough to use.

**Comment by mitchcapper** [comment]
I thought about modifying the current event, I would consider that a breaking change however as if you currently count on it to only fire on first bind it now is called more often.  I am happy with that solution though not sure about others although at 63 is fairly new it may not be a big deal.

ObjectBoundInJavascript is fine I think.  In JS users may call it once or over multiple lines.

I agree allowing users to control when object bind is useful, especially as if they do want autobinding they can implement something like OnContextCreated.

The main goal of all of this was just to give a way from c# to make sure something is bound.  If there is a new function or existing function with args that is always called is greatly simplifies a helper class to something like the following:  [Helper.cs](https://github.com/cefsharp/CefSharp/files/1913647/Helper.cs.txt)

**Comment by mitchcapper** [comment]
I would vote for first on a method that takes an open ended array of args.  Any of these changes sound great.

**Comment by amaitland** [comment]
Marking this as `blocking` as I think this should happen before `65` is released. I'll try and get it sorted next week.

**Comment by amaitland** [comment]
A rough implementation was merged into `master` with https://github.com/cefsharp/CefSharp/commit/8f7692ae6e5c75e26accf49d57f5d8460c81b271

TODO
- [ ] Remove code duplication
- [ ] Add more qunit tests

As it's not ready for wide spread release yet, I've merged the breaking changes into the `cefsharp/65` branch with commit https://github.com/cefsharp/CefSharp/commit/801d53b3cf65da168bdd1cd9d7316b4a25f733b9

Will look at releasing a `65.0.0-pre01` set of packages very shortly.

Let me know what you think. You can see the new `JS` syntax at https://github.com/cefsharp/CefSharp/commit/8f7692ae6e5c75e26accf49d57f5d8460c81b271#diff-f99ba049add937c58daa5a3d4533da17R473

**Comment by mitchcapper** [comment]
Syntax and event args seem easy to follow will certainly simplify binding resolution from .net as well.

**Comment by amaitland** [comment]
Suggestions on naming welcome 😃

**Comment by amaitland** [comment]
> The naming seems straightforward enough and while it is easy to use them as is I do wonder if NotifyIfAlreadyBound isn't better done as an option from the .net side rather than JS

From `JS` you can change the param when you call it, from `.Net` if the objects are cached in the render process then it wouldn't be possible to change the notify option. The notification is triggered in the render process. Maybe it's not something that needs to be changed once set? I was thinking all new options going froward would be in `JS` unless there's a really strong case for them to be in `.Net`. I was going to look at adding a `parent` param, so instead of binding directly to window, you can provide an object that becomes the parent. I'm sure there will be other params for use cases that I'm not aware of yet.

> NotifyEvenIfAlreadyBound, AlwaysGenerateEvent, AlwaysGenerateNotification, ForceNotify if you are looking for other options.

Probably just go with the current name unless there's a strong opinion for something else.

**Comment by amaitland** [comment]
Have added `EnsureObjectBoundAsync` extension method, there's a new event called `IJavascriptObjectRepository.ObjectsBoundInJavascript` that simplifies the code greatly (hopefully the name isn't confusing, it's simple called a single time when the object bound message is received.

Example looks something like:

```c#
browser.EnsureObjectBoundAsync("bound").ContinueWith(t =>
{
  var boundObjects = t.Result;
});
```

Obviously you can `await` the `Task`. Still needs a bit more testing/cleanup. It doesn't check `browser.CanExecuteJavascriptInMainFrame` like the example provided, there's still issue #2250 that makes using that unreliable after a process switch. Let me know what you think.

**Comment by mitchcapper** [comment]
ObjectsBoundInJavascript  certainly simplifies it and as a user can use either shouldn't make it too difficult for users.

EnsureObjectBound helper will likely be useful for others so should make a nice addition for transitioning existing code.

**Comment by amaitland** [comment]
Section added to the `General Usage Guide` see https://github.com/cefsharp/CefSharp/wiki/General-Usage/_compare/f3ded5d15cb167569c62bf4134f3bb1b049764e9...72cdf2f2d5666db7cfac100f35494c54a0562755

I'm not totally happy with the formatting, overall the whole `JSB` section could use some work, there are quite a few options now. Open to suggestions for how to improve the formatting and make things clearer.

**Comment by amaitland** [comment]
`65.0.0` should appear on `Nuget.org` shortly.

The documentation still needs some work, so I've creating #2455
